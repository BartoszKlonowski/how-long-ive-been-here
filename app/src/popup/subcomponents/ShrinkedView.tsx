import React, {useEffect, useState} from "react";
import {getActiveTabDomainFromURL, getHours, getMinutes, getSeconds, getWebsiteIconObject, Icon} from "../Utils";
import browser from "webextension-polyfill";
import Database from "../../engine/Database";

const getCurrentTimeForCurrentDomain = (domain: string): number => {
    const db = new Database();
    const timeSpent = (db.readTimeSpent(domain) as number) || 0;
    return timeSpent;
};

export const ShrinkedView = () => {
    const [icon, setIcon] = useState<Icon>(getWebsiteIconObject(""));
    const [timeInSeconds, setTimeInSeconds] = useState(0);
    const [activeDomain, setActiveDomain] = useState("");

    useEffect(() => {
        browser.tabs
            .query({active: true})
            .then((result) => {
                const domain = getActiveTabDomainFromURL(result[0].url!) || "";
                setIcon(getWebsiteIconObject(domain));
                setTimeInSeconds(getCurrentTimeForCurrentDomain(domain));
                setActiveDomain(domain);
            })
            .catch((error: Error) => {
                console.error(error.message);
            });
    }, []);

    useEffect(() => {
        return () => {
            const db = new Database();
            db.writeTimeSpent(activeDomain, timeInSeconds);
        };
    }, [activeDomain, timeInSeconds]);

    useEffect(() => {
        const currentTimer = setInterval(() => {
            setTimeInSeconds(timeInSeconds + 1);
        }, 1000);

        return () => clearInterval(currentTimer);
    });

    return (
        <div className="shrinked-view-container">
            <div className="shrinked-view-icon-container">
                <img className="shrinked-view-icon" width={icon.size} height={icon.size} src={icon.src} />
            </div>
            <div className="shrinked-view-time-spent-text-container">
                <div className="shrinked-view-time-spent-text">
                    {getHours(timeInSeconds)}:{getMinutes(timeInSeconds)}:{getSeconds(timeInSeconds)}
                </div>
            </div>
        </div>
    );
};

export default ShrinkedView;
