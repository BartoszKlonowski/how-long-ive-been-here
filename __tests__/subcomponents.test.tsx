import React from "react";
import TestRenderer, {act, ReactTestInstance, ReactTestRenderer} from "react-test-renderer";
import ShrinkedView from "../app/src/popup/subcomponents/ShrinkedView";

function render(): ReactTestRenderer {
    const shrinkedView = TestRenderer.create(<ShrinkedView />);
    return shrinkedView;
}

async function renderAsObject(): Promise<ReactTestInstance> {
    let component;
    await act(async () => {
        component = TestRenderer.create(<ShrinkedView />);
    });
    return JSON.parse(JSON.stringify(component.toJSON()));
}

function getChild(renderedObject: ReactTestInstance, childIndex: number): ReactTestInstance {
    const child = renderedObject.children[childIndex] as ReactTestInstance;
    return child;
}

describe("ShrinkedView", () => {
    it("renders correctly according to snapshot", () => {
        const shrinkedView = render();
        expect(shrinkedView.toJSON()).toMatchSnapshot();
    });

    it("renders icon as image with default src when missing icon on website", async () => {
        global.browser.tabs.query = () => {
            return new Promise((resolve) => {
                resolve([{url: ""}]);
            });
        };
        const shrinkedView = await renderAsObject();
        const img = getChild(getChild(shrinkedView, 0), 0);
        expect(img).toBeDefined();
        expect(img.type).toBe("img");
        expect(img.props.src).toContain("../resources/missing-website-favicon.png");
    });

    it("renders icon as image with proper src when icon is available on website", async () => {
        global.browser.tabs.query = () => {
            return new Promise((resolve) => {
                resolve([{url: "proper-existing-icon-url"}]);
            });
        };
        const shrinkedView = await renderAsObject();
        const img = getChild(getChild(shrinkedView, 0), 0);
        expect(img).toBeDefined();
        expect(img.type).toBe("img");
        expect(img.props.src).toContain("http://www.google.com/s2/favicons?domain=proper-existing-icon-url");
    });

    it("displays 0 time spent for incorrect domain", async () => {
        global.browser.tabs.query = () => {
            return new Promise((resolve) => {
                resolve([{url: ""}]);
            });
        };
        const shrinkedView = await renderAsObject();
        const timeSpentText = getChild(getChild(shrinkedView, 1), 0);
        console.log(JSON.stringify(timeSpentText));
        expect(timeSpentText).toBeDefined();
        expect(`${getChild(timeSpentText, 0)}:${getChild(timeSpentText, 2)}:${getChild(timeSpentText, 4)}`).toBe(
            "0:00:00"
        );
    });
});
