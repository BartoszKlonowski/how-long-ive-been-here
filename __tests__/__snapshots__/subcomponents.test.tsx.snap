// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpandedView renders correctly according to snapshot 1`] = `
<div
  className="expanded-view-list-container"
>
  <ul
    className="expanded-view-list"
  >
    <li>
      <div
        className="expanded-view-tile-container"
      >
        <div
          className="expanded-view-icon-container"
        >
          <img
            className="expanded-view-icon"
            height={20}
            src="https://icons.duckduckgo.com/ip3/w3schools.com.ico"
            width={20}
          />
        </div>
        <div
          className="expanded-view-domain-container"
        >
          <div
            className="expanded-view-domain-text"
          >
            w3schools.com
          </div>
        </div>
        <div
          className="expanded-view-time-spent-text-container"
        >
          <div
            className="expanded-view-time-spent-text"
          >
            0
            :
            01
            :
            13
          </div>
        </div>
      </div>
    </li>
    <li>
      <div
        className="expanded-view-tile-container"
      >
        <div
          className="expanded-view-icon-container"
        >
          <img
            className="expanded-view-icon"
            height={20}
            src="https://icons.duckduckgo.com/ip3/ackoverflow.com.ico"
            width={20}
          />
        </div>
        <div
          className="expanded-view-domain-container"
        >
          <div
            className="expanded-view-domain-text"
          >
            ackoverflow.com
          </div>
        </div>
        <div
          className="expanded-view-time-spent-text-container"
        >
          <div
            className="expanded-view-time-spent-text"
          >
            0
            :
            11
            :
            14
          </div>
        </div>
      </div>
    </li>
    <li>
      <div
        className="expanded-view-tile-container"
      >
        <div
          className="expanded-view-icon-container"
        >
          <img
            className="expanded-view-icon"
            height={20}
            src="https://icons.duckduckgo.com/ip3/github.com.ico"
            width={20}
          />
        </div>
        <div
          className="expanded-view-domain-container"
        >
          <div
            className="expanded-view-domain-text"
          >
            github.com
          </div>
        </div>
        <div
          className="expanded-view-time-spent-text-container"
        >
          <div
            className="expanded-view-time-spent-text"
          >
            0
            :
            10
            :
            07
          </div>
        </div>
      </div>
    </li>
    <li>
      <div
        className="expanded-view-tile-container"
      >
        <div
          className="expanded-view-icon-container"
        >
          <img
            className="expanded-view-icon"
            height={20}
            src="https://icons.duckduckgo.com/ip3/google.com.ico"
            width={20}
          />
        </div>
        <div
          className="expanded-view-domain-container"
        >
          <div
            className="expanded-view-domain-text"
          >
            google.com
          </div>
        </div>
        <div
          className="expanded-view-time-spent-text-container"
        >
          <div
            className="expanded-view-time-spent-text"
          >
            0
            :
            00
            :
            07
          </div>
        </div>
      </div>
    </li>
    <li>
      <div
        className="expanded-view-tile-container"
      >
        <div
          className="expanded-view-icon-container"
        >
          <img
            className="expanded-view-icon"
            height={20}
            src="https://icons.duckduckgo.com/ip3/npmjs.com.ico"
            width={20}
          />
        </div>
        <div
          className="expanded-view-domain-container"
        >
          <div
            className="expanded-view-domain-text"
          >
            npmjs.com
          </div>
        </div>
        <div
          className="expanded-view-time-spent-text-container"
        >
          <div
            className="expanded-view-time-spent-text"
          >
            0
            :
            00
            :
            05
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`ShrinkedView renders correctly according to snapshot 1`] = `
<div
  className="shrinked-view-container"
>
  <div
    className="shrinked-view-icon-container"
  >
    <img
      className="shrinked-view-icon"
      height={20}
      src="../resources/missing-website-favicon.png"
      width={20}
    />
  </div>
  <div
    className="shrinked-view-time-spent-text-container"
  >
    <div
      className="shrinked-view-time-spent-text"
    >
      0
      :
      00
      :
      00
    </div>
  </div>
</div>
`;

exports[`ViewChangeButton renders correctly regarding the snapshot 1`] = `
<button
  className="view-change-button"
  itemType="button"
  onClick={[Function]}
>
  <div
    className="view-change-button-text"
  >
    More
  </div>
</button>
`;
